assets/app/view/game/buy_sell_shares.rb

```diff
@@ -12,114 +13,138 @@ class BuySellShares < Snabberb::Component
-              "Buy #{@game.class::IPO_NAME} Share",
-            )
-          end
+        children << h(SellShares, player: @current_entity, corporation: @corporation)
 
-          # Put up one buy button for each buyable percentage share type in market.
-          # In case there are more than one type of percentages in market or if shares are not the
-          # standard percent (e.g. 5% in 18MEX), show percentage type on button.
-          # Do skip president's share in case there are other shares available.
-          buyables = pool_shares
-            .select { |share| step.can_buy?(current_entity, share) }
-            .reject { |share| share.president && pool_shares.size > 1 }
-          buyables.each do |share|
-            text = buyables.size > 1 || share.percent != @corporation.share_percent ? "#{share.percent}% " : ''
-            children << h(:button,
-                          { on: { click: -> { buy_share(current_entity, share) } } },
-                          "Buy #{text}Market Share")
-          end
+        h(:div, children.compact)
+      end
 
-          if price_protection
-            children << h(
-              :button,
-              { on: { click: -> { buy_share(current_entity, price_protection.shares) } } },
-              'Protect Shares',
-            )
-          end
+      def render_buy_shares
+        return [] unless @step.current_actions.include?('buy_shares')
 
-          if ipo_share && (swap_share = step.swap_buy(current_entity, @corporation, ipo_share))
-            puts("ipo_share = #{ipo_share.class} with price #{ipo_share.price}")
-            reduced_price = @game.format_currency(ipo_share.price - swap_share.price)
-            children << h(
-              :button,
-              { on: { click: -> { buy_share(current_entity, ipo_share, swap: swap_share) } } },
-              "Buy #{@game.class::IPO_NAME} Share (#{reduced_price} + #{swap_share.percent}% Share)",
-            )
-          end
+        children = []
 
-          pool_shares.each do |pool_share|
-            next unless (swap_share = step.swap_buy(current_entity, @corporation, pool_share))
+        children.concat(render_ipo_shares)
+        children.concat(render_market_shares)
+        children.concat(render_price_protection)
+        children.concat(render_reduced_price_shares(@ipo_shares, source: @game.class::IPO_NAME))
+        children.concat(render_reduced_price_shares(@pool_shares))
 
-            puts("pool_share = #{pool_share.class} with price #{pool_share.price}")
-            reduced_price = @game.format_currency(pool_share.price - swap_share.price)
-            children << h(
-              :button,
-              { on: { click: -> { buy_share(current_entity, pool_share, swap: swap_share) } } },
-              "Buy #{pool_share.percent}% Market Share (#{reduced_price} + #{swap_share.percent}% Share)",
-            )
-          end
+        children.compact
```

Comment:
do you need to call compact if these just all return arrays?

Reply:
Good point, might as well just do a single compact at the end.